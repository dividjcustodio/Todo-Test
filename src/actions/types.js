export const CHANGE_USER_DATA = "CHANGE_USER_DATA";
export const CHANGE_LIST_DATA = "CHANGE_LIST_DATA";
