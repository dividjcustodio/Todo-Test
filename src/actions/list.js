import { CHANGE_LIST_DATA } from "./types";
export const getDataStart = dispatch => {
  fetch(`https://jsonplaceholder.typicode.com/todos`, {
    method: "GET",
    mode: "cors",
    cache: "no-cache",
    credentials: "same-origin",
    headers: new Headers(),
    redirect: "follow",
    referrer: "no-referrer"
  })
    .then(response => {
      return response.json();
    })
    .then(data => {
      dispatch(getDataSuccess(data));
    });
};
export const getDataSuccess = data => {
  return {
    type: CHANGE_LIST_DATA,
    data: { data }
  };
};
export const getDataDetailStart = async (id, dataSet) => {
  const res = await fetch(`https://jsonplaceholder.typicode.com/todos/${id}`, {
    method: "GET",
    mode: "cors",
    cache: "no-cache",
    credentials: "same-origin",
    headers: new Headers(),
    redirect: "follow",
    referrer: "no-referrer"
  });
  const data = await res.json();
  dataSet(data);
};
