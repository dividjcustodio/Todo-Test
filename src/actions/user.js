import { CHANGE_USER_DATA } from "./types";

export const signInAction = data => {
  window.sessionStorage.setItem("user", JSON.stringify(data));

  return {
    data: { ...data, isAuthenticated: true },
    type: CHANGE_USER_DATA
  };
};

export const logout = (data, history) => {
  window.sessionStorage.setItem("user", "");

  return {
    data: { name: "", password: "", isAuthenticated: false },
    type: CHANGE_USER_DATA
  };
};
