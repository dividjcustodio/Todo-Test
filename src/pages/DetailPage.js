import React from "react";
import { getDataDetailStart } from "../actions/list";
import "../App.css";
import { Container, Card  } from '@material-ui/core';
import useStyles from "../assets/styles";

export function DetailPage({ match }) {
  const classes = useStyles();
  const id = Number.isInteger(+match.params.id) ? match.params.id : 0;

  const [data, dataSet] = React.useState([]);
  React.useEffect(() => {
    getDataDetailStart(id, dataSet);
  }, []);

  return (
    <Container maxWidth="sm" className={classes.signInPage}>
      <Card className={`${classes.w100} ${classes.p50}`}>
        <h2>Title : { data.title }</h2>
        <p>UserId : {data.userId}</p>
        <p>ID : {data.id}</p>
        <p>State : {data.completed ? 
          <span className={classes.greenFont}>completed</span> : 
          <span className={classes.redFont}>no completed</span>
        }</p>
      </Card>
    </Container>
  );
}

export default DetailPage;
