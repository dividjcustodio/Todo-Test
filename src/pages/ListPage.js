import React from "react";
import { connect } from "react-redux";
import { List, Container  } from '@material-ui/core';
import { getDataStart } from "../actions/list";

import useStyles from "../assets/styles";
import Item from "../components/Item";

export function ListPage({ data, getData }) {
  const classes = useStyles();

  React.useEffect(() => {
    getData();
  }, []);
  return (
  <Container maxWidth="sm">
    <h2 className={`${classes.displayCenter} ${classes.w100}`}>List</h2>
    <List>
      {data.map(item => (
        <Item {...item} key={item.id}/>
      ))}
    </List>
  </Container>
  );
}
const mapStateToProps = store => ({
  data: store.list.data
});
const dispatchToProps = dispatch => {
  return {
    getData: () => dispatch(getDataStart())
  };
};

export default connect(mapStateToProps, dispatchToProps)(ListPage);
