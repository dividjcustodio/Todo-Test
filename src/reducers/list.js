import { CHANGE_LIST_DATA } from "../actions/types";

const INITIAL_STATE = {
  data: []
};

export default (state = INITIAL_STATE, action) => {
  switch (action.type) {
    case CHANGE_LIST_DATA:
      return {
        ...state,
        ...action.data
      };

    default:
      return state;
  }
};
