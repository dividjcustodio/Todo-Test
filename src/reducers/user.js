import { CHANGE_USER_DATA } from "../actions/types";

const INITIAL_STATE = {
  isAuthenticated: false,
  name: "",
  password: ""
};

export default (state = INITIAL_STATE, action) => {
  switch (action.type) {
    case CHANGE_USER_DATA:
      return {
        ...state,
        ...action.data
      };

    default:
      return state;
  }
};
