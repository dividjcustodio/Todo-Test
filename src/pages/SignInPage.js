import React from "react";
import { connect } from "react-redux";
import { Button, Container, Card, TextField  } from '@material-ui/core';
import { signInAction } from "../actions/user";
import useStyles from "../assets/styles";

function SignInPage({ signInAction, history }) {
  const classes = useStyles();
  const [data, dataSet] = React.useState({
    name: "",
    password: ""
  });
  const handleChangeValue = event => {
    const { name, value } = event.target;

    dataSet(prev => ({
      ...prev,
      [name]: value
    }));
  };
  const onSubmit = () => {
    if (data.name) {
      signInAction(data, history);
    }
  };
  return (
    <Container maxWidth="sm" className={classes.signInPage}>
      <Card className={classes.w100}>
        <h3 className={classes.textAlignCenter}>Sign In</h3>
        <div className={ `${classes.w100} ${classes.displayCenter} ${classes.mb50}`}>
          <TextField
            required 
            onChange={handleChangeValue}
            value={data.name}
            name="name"
            type="text"
            id="filled-basic"
            label="UserName"
          />
        </div>
        <div  className={ `${classes.w100} ${classes.displayCenter} ${classes.mb50}`}>
          <Button variant="contained" color="primary" onClick={onSubmit}>submit</Button>
        </div>
      </Card>
    </Container>
  );
}
export default connect(null, { signInAction })(SignInPage);
