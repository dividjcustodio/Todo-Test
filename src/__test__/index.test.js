import React from "react";
import nock from "nock";
import Enzyme, { mount, configure } from "enzyme";
import ConnectedApp, { App } from "../App";
import { SignInPage } from "../pages/SignInPage";
import { ListPage } from "../pages/ListPage";
import DetailPage from "../pages/DetailPage";
import { MemoryRouter } from 'react-router'
import configureStore from "redux-mock-store";
import { Provider } from "react-redux";
import Adapter from "enzyme-adapter-react-16";
import { logout, signInAction } from "../actions/user";
import { getDataSuccess, getDataStart } from "../actions/list";
import { item, list } from "../helpers/mock";
import thunk from "redux-thunk";
import { createShallow, createMount } from '@material-ui/core/test-utils';

configure({ adapter: new Adapter() });
Enzyme.configure({ adapter: new Adapter() });
let userData = {
  isAuthenticated: false
};

const dataItem = item;
const dataList = list;

let dataLocal = localStorage.getItem("user");
if (dataLocal) {
  userData = {
    isAuthenticated: true,
    ...JSON.parse(dataLocal)
  };
}
const initialS = { user: { ...userData }, list: { data: [] } };

// Snapshot for SignInPage React Component


describe('>>>Pages --- Snapshot', () => {
  const shallow = createShallow();
  const mockStore = configureStore();
  let store = mockStore(initialS);

  const getData = jest.fn();
  const ListPageContainer = shallow(
    <MemoryRouter>
      <ListPage data={[dataItem]} getData={getData} />
    </MemoryRouter>
  )
  const matchParam = {
    params: {
      id: 3
    }
  }
  const detailPageContainer = shallow(<MemoryRouter><DetailPage match={matchParam}/></MemoryRouter>);
  const SignInPageContainer = shallow(<MemoryRouter><SignInPage/></MemoryRouter>);

  it("+++capturing Snapshot of ListPage", () => {
    expect(ListPageContainer.html()).toMatchSnapshot();
  });
  it("+++capturing Snapshot of DetailPage", () => {
    expect(detailPageContainer.html()).toMatchSnapshot();
  });
  it("+++capturing Snapshot of SignInPage", () => {
    expect(SignInPageContainer.html()).toMatchSnapshot();
  });

});


//*******************************************************************************************************
describe(">>>H O M E --- Shallow Render REACT COMPONENTS", () => {
  const shallow = createShallow();
  const mockStore = configureStore();
  const store = mockStore(initialS);

  const props = {
    user: { ...userData }
  };

  const wrapper1 = shallow(<ConnectedApp store={store} {...props} />);

  it("+++ render the DUMB component", () => {
    expect(wrapper1.length).toEqual(1);
  });
});

//*******************************************************************************************************
describe(">>>H O M E --- REACT-REDUX (Shallow + passing the {store} directly)", () => {
  const shallow = createShallow();
  const initialState = initialS;
  const mockStore = configureStore();
  const store = mockStore(initialState);

  const container = shallow(<ConnectedApp store={store} />);
  it("+++ render the connected(SMART) component", () => {
    expect(container.length).toEqual(1);
  });
});

//*******************************************************************************************************
describe(">>>H O M E --- REACT-REDUX (Mount + wrapping in <Provider>)", () => {
  const initialState = initialS;
  const middlewares = [thunk];
  const mockStore = configureStore(middlewares);
  const mount = createMount();
  const store = mockStore(initialState);
  let fetchTodosData = dataList;
  const wrapper = mount(
    <Provider store={store}>
      <ConnectedApp />
    </Provider>
  );
  afterEach(() => {
    // clear all HTTP mocks after each test
    nock.cleanAll();
  });
  it("+++ render the connected(SMART) component", () => {
    expect(wrapper.find(ConnectedApp).length).toEqual(1);
  });

  it("+++ check action user ", () => {
    let action;
    store.dispatch(signInAction({ name: "some", password: "2323" }));
    store.dispatch(logout());
    action = store.getActions();
    expect(action[0].type).toBe("CHANGE_USER_DATA");
    expect(action[1].type).toBe("CHANGE_USER_DATA");
  });

  it("+++ check action list", () => {
    // Dispatch the createSuccess action with the values of a new to-do.
    store.dispatch(getDataSuccess([dataItem]));
    expect(store.getActions()).toMatchSnapshot();
  });

  it("creates async check action list", () => {
    // Simulate a successful response
    nock("http://jsonplaceholder.typicode.com")
      .defaultReplyHeaders({ "access-control-allow-origin": "*" })
      .get("/todos") // Route to catch and mock
      .reply(200, fetchTodosData); // Mock reponse code and data
    // Dispatch action to fetch to-dos
    store.dispatch(getDataStart()).then(() => {
      // return of async actions
      expect(store.getActions()).toMatchSnapshot();
    });
  });
});
