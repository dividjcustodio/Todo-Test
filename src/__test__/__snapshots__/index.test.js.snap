// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`>>>H O M E --- REACT-REDUX (Mount + wrapping in <Provider>) +++ check action list 1`] = `
Array [
  Object {
    "data": Object {
      "isAuthenticated": true,
      "name": "some",
      "password": "2323",
    },
    "type": "CHANGE_USER_DATA",
  },
  Object {
    "data": Object {
      "isAuthenticated": false,
      "name": "",
      "password": "",
    },
    "type": "CHANGE_USER_DATA",
  },
  Object {
    "data": Object {
      "data": Array [
        Object {
          "completed": false,
          "id": 1,
          "title": "delectus aut autem",
          "userId": 1,
        },
      ],
    },
    "type": "CHANGE_LIST_DATA",
  },
]
`;

exports[`>>>Pages --- Snapshot +++capturing Snapshot of DetailPage 1`] = `"<div class=\\"MuiContainer-root makeStyles-signInPage-1 MuiContainer-maxWidthSm\\"><div class=\\"MuiPaper-root MuiCard-root makeStyles-w100-4 makeStyles-p50-6 MuiPaper-elevation1 MuiPaper-rounded\\"><h2>Title : </h2><p>UserId : </p><p>ID : </p><p>State : <span class=\\"makeStyles-redFont-8\\">no completed</span></p></div></div>"`;

exports[`>>>Pages --- Snapshot +++capturing Snapshot of ListPage 1`] = `"<div class=\\"MuiContainer-root MuiContainer-maxWidthSm\\"><h2 class=\\"makeStyles-displayCenter-2 makeStyles-w100-4\\">List</h2><ul class=\\"MuiList-root MuiList-padding\\"><li class=\\"MuiListItem-root MuiListItem-gutters\\" data-testid=\\"item\\"><a href=\\"/detail/1\\">delectus aut autem</a><div class=\\"MuiListItemText-root makeStyles-redFont-8 makeStyles-flexEnd-3\\"><span class=\\"MuiTypography-root MuiListItemText-primary MuiTypography-body1\\">no completed</span></div></li></ul></div>"`;

exports[`>>>Pages --- Snapshot +++capturing Snapshot of SignInPage 1`] = `"<div class=\\"MuiContainer-root makeStyles-signInPage-1 MuiContainer-maxWidthSm\\"><div class=\\"MuiPaper-root MuiCard-root makeStyles-w100-4 MuiPaper-elevation1 MuiPaper-rounded\\"><h3 class=\\"makeStyles-textAlignCenter-7\\">Sign In</h3><div class=\\"makeStyles-w100-4 makeStyles-displayCenter-2 makeStyles-mb50-5\\"><div class=\\"MuiFormControl-root MuiTextField-root\\"><label class=\\"MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated Mui-required Mui-required\\" data-shrink=\\"false\\" for=\\"filled-basic\\" id=\\"filled-basic-label\\">UserName<span class=\\"MuiFormLabel-asterisk MuiInputLabel-asterisk\\">â€‰*</span></label><div class=\\"MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl\\"><input type=\\"text\\" aria-invalid=\\"false\\" id=\\"filled-basic\\" name=\\"name\\" required=\\"\\" value=\\"\\" class=\\"MuiInputBase-input MuiInput-input\\"/></div></div></div><div class=\\"makeStyles-w100-4 makeStyles-displayCenter-2 makeStyles-mb50-5\\"><button class=\\"MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary\\" tabindex=\\"0\\" type=\\"button\\"><span class=\\"MuiButton-label\\">submit</span></button></div></div></div>"`;
